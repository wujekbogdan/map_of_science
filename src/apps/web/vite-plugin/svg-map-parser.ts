import { createFilter } from "@rollup/pluginutils";
import { readFileSync, existsSync } from "fs";
import camelCase from "lodash/camelCase";
import { resolve } from "path";
import styleToObject from "style-to-object";
import { svgPathBbox } from "svg-path-bbox";
import { Plugin } from "vite";
import { Parser } from "xml2js";
import { z } from "zod";

const l1L2Schema = z.object({
  attributes: z.object({
    "inkscape:label": z.string(),
    id: z.string(),
    style: z.string(),
  }),
  path: z.array(
    z.object({
      attributes: z.object({
        class: z.string().optional(),
        "inkscape:label": z.string(),
        style: z.string(),
        d: z.string(),
        id: z.string(),
      }),
    }),
  ),
});

const l3Schema = z.object({
  attributes: z.object({
    "inkscape:label": z.string(),
    id: z.string(),
    style: z.string(),
  }),
  g: z.array(
    z.object({
      attributes: z.object({
        "inkscape:label": z.string(),
        id: z.string(),
      }),
      rect: z.array(
        z.object({
          attributes: z.object({
            id: z.string(),
            style: z.string(),
            width: z.coerce.number(),
            height: z.coerce.number(),
            x: z.coerce.number(),
            y: z.coerce.number(),
            "inkscape:label": z.string(),
          }),
        }),
      ),
    }),
  ),
});

const schema = z.object({
  svg: z.object({
    g: z.tuple([l1L2Schema, l1L2Schema, l3Schema, z.unknown()]),
  }),
});

type ParsedSvg = z.infer<typeof schema>;

export const parse = async (svgString: string) => {
  // Type casting is fine here because we're validating the parsed object with zod anyway
  const parser = new Parser({
    attrkey: "attributes",
  });
  const parsedSvg = (await parser.parseStringPromise(svgString)) as ParsedSvg;

  const validated = schema.parse(parsedSvg);

  const style = (style: string) => {
    return Object.fromEntries(
      Object.entries(styleToObject(style) ?? {}).map(([property, value]) => [
        camelCase(property),
        value,
      ]),
    );
  };

  const map1st2ndLayer = (layer: z.infer<typeof l1L2Schema>) => ({
    attributes: {
      id: layer.attributes.id,
      label: layer.attributes["inkscape:label"],
      style: style(layer.attributes.style),
    },
    children: layer.path.map(({ attributes }) => ({
      path: {
        id: attributes.id,
        label: attributes["inkscape:label"],
        style: style(attributes.style),
        d: attributes.d,
        boundingBox: (() => {
          const [minx, miny, maxX, maxY] = svgPathBbox(attributes.d);
          return {
            min: { x: minx, y: miny },
            max: { x: maxX, y: maxY },
            center: {
              x: (minx + maxX) / 2,
              y: (miny + maxY) / 2,
            },
          };
        })(),
      },
    })),
  });

  const map3rdLayer = (layer: z.infer<typeof l3Schema>) => ({
    attributes: {
      id: layer.attributes.id,
      label: layer.attributes["inkscape:label"],
      style: style(layer.attributes.style),
    },
    groups: layer.g.map((group) => ({
      attributes: {
        id: group.attributes.id,
        label: group.attributes["inkscape:label"],
      },
      children: group.rect.map(({ attributes }) => ({
        rect: {
          id: attributes.id,
          label: attributes["inkscape:label"],
          style: style(attributes.style),
          width: attributes.width,
          height: attributes.height,
          x: attributes.x,
          y: attributes.y,
          boundingBox: {
            min: { x: attributes.x, y: attributes.y },
            max: {
              x: attributes.x + attributes.width,
              y: attributes.y + attributes.height,
            },
            center: {
              x: attributes.x + attributes.width / 2,
              y: attributes.y + attributes.height / 2,
            },
          },
        },
      })),
    })),
  });

  const [layer1, layer2, layer3] = validated.svg.g;

  return {
    layer1: map1st2ndLayer(layer1),
    layer2: map1st2ndLayer(layer2),
    layer3: map3rdLayer(layer3),
  };
};

export type MapSvgRepresentation = Awaited<ReturnType<typeof parse>>;

export default function svgMapParser(): Plugin {
  const filter = createFilter(/\.(svg)$/, null);

  return {
    name: "vite-plugin-svg-map-parser",

    async transform(_, id) {
      if (!id.includes("?parse")) return null;

      const [filepath] = id.split("?"); // Get the actual file path
      if (!filter(filepath)) return null;

      if (!existsSync(filepath)) {
        this.error(`File not found: ${filepath}`);
      }

      const svgData = readFileSync(resolve(filepath), "utf-8");

      return {
        code: `export default ${JSON.stringify(await parse(svgData))};`,
        map: null,
      };
    },
  };
}
