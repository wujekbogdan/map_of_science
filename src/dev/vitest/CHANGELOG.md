# @map-of-science/vitest

## 0.1.0

### Minor Changes

- e86d99f: Extract the Vitest integration into a separate `@map-of-science/vitest` package.
