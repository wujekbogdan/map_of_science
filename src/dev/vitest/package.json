{
  "name": "@map-of-science/vitest",
  "version": "0.1.0",
  "private": true,
  "type": "module",
  "keywords": [],
  "author": "",
  "license": "ISC",
  "exports": "./src/index.ts",
  "dependencies": {
    "vitest": "^3.1.3"
  },
  "bin": {
    "vitest": "./node_modules/vitest/vitest.mjs"
  }
}
