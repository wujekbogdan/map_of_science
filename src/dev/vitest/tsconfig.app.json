{
  "compilerOptions": {
    "tsBuildInfoFile": "./node_modules/.tmp/tsconfig.app.tsbuildinfo",
    "target": "ES2020",
    "useDefineForClassFields": true,
    "lib": ["ES2020", "DOM", "DOM.Iterable"],
    "module": "ESNext",
    "skipLibCheck": true,

    /* Bundler mode */
    "moduleResolution": "Bundler",
    "allowImportingTsExtensions": true,
    "isolatedModules": true,
    "moduleDetection": "force",
    "noEmit": true,
    "jsx": "react-jsx",

    /* Linting */
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedSideEffectImports": true,

    // TODO: Remove once the project is fully migrated to TypeScript
    "allowJs": true,
    "checkJs": false,
    "allowSyntheticDefaultImports": true
  },
  "include": ["src"]
}
