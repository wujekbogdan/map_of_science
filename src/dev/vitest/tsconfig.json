{
  "extends": "@map-of-science/typescript/node"
}