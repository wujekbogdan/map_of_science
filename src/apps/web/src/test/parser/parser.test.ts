import { describe, it, expect } from "vitest";
import { ZodError } from "zod";
import { parse } from "../../../vite-plugin/svg-map-parser.ts";
import validSvg from "./valid-svg.test.svg?raw";

// TODO: Move this test to /vite-plugin/parser.test.ts
// Temporarily placed here because, for some reason, even when Vite test is explicitly configured to include that file, it still doesn't work
describe("SVG Map Parser Vite Plugin", () => {
  it("should parse a valid SVG", async () => {
    expect(await parse(validSvg)).toMatchSnapshot();
  });

  it("should throw an error if the provided string isn't a valid SVG", async () => {
    await expect(parse("invalid svg")).rejects.toThrow(
      "Non-whitespace before first tag",
    );
  });

  it("should throw a schema validation error if the provided SVG is valid but does not follow the schema", async () => {
    const svg = `
      <svg>
        <g>
          <rect />
        </g>
      </svg>
    `;
    await expect(parse(svg)).rejects.toThrow(ZodError);
  });
});
