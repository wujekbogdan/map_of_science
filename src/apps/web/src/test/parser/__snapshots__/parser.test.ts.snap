// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SVG Map Parser Vite Plugin > should parse a valid SVG 1`] = `
{
  "layer1": {
    "attributes": {
      "id": "layer1",
      "label": "L1",
      "style": {
        "display": "inline",
      },
    },
    "children": [
      {
        "path": {
          "boundingBox": {
            "center": {
              "x": -431.8910662922816,
              "y": 22.71598460951293,
            },
            "max": {
              "x": -322.8051709687215,
              "y": 258.8283511316347,
            },
            "min": {
              "x": -540.9769616158417,
              "y": -213.39638191260883,
            },
          },
          "d": "m -513.94556,55.008741 c -3.96216,-8.3378 -34.16543,-44.65728 -25.46517,-102.3627 8.70025,-57.705411 5.86342,-101.550601 37.68697,-131.820851 31.82354,-30.27024 40.89514,-35.91347 60.06968,-33.83659 19.17454,2.07688 46.32195,5.66537 59.97734,31.71243 13.65538,26.04706 27.93241,67.40747 33.41685,91.411579 5.98485,26.194261 0.72002,75.858629 -1.77023,123.183841 -2.55731,48.599759 30.37753,116.49709 26.97856,162.43643 -1.64111,22.18064 4.44665,45.28185 -10.37048,56.7905 -10.54822,8.19294 -22.64597,15.62258 -50.50341,-27.15057 C -450.33278,123.40902 -513.94556,55.008741 -513.94556,55.008741 Z",
          "id": "path186",
          "label": "#Symbole-bity-sygnały",
          "style": {
            "clipRule": "evenodd",
            "display": "inline",
            "fill": "#a3d293",
            "fillOpacity": "1",
            "fillRule": "evenodd",
            "imageRendering": "optimizeQuality",
            "paintOrder": "stroke fill markers",
            "shapeRendering": "geometricPrecision",
            "stroke": "none",
            "strokeDasharray": "none",
            "strokeLinecap": "round",
            "strokeLinejoin": "round",
            "strokeOpacity": "1",
            "strokeWidth": "1",
            "textRendering": "geometricPrecision",
          },
        },
      },
      {
        "path": {
          "boundingBox": {
            "center": {
              "x": -175.98698476092548,
              "y": 40.46505714629674,
            },
            "max": {
              "x": -4.789039521850984,
              "y": 265.2055539220782,
            },
            "min": {
              "x": -347.18492999999995,
              "y": -184.2754396294847,
            },
          },
          "d": "m -331.82381,3.0279384 c 1.38625,-37.9455014 -27.47386,-116.3727984 30.36309,-166.3343884 l 25.01454,-18.23459 c 18.29324,-7.85022 53.77387,2.95113 74.34538,10.43231 l 48.52124,39.04205 69.825524,35.157208 c 0,0 7.75298,8.692 14.21049,17.53389 6.4575,8.84189 13.97315,19.24089 19.91458,33.67001 5.94142,14.42912 9.69933,29.25972 11.92158,37.4621304 2.22225,8.20241005 4.46437,22.3314096 7.53552,35.3024196 3.07114,12.97101 2.22242,14.25625 3.7579,26.62886 1.53549,12.37261 3.35065,24.35691 4.03741,34.17579 0.68676,9.81888 22.11143006,41.848202 16.7176,61.367762 -5.39383,19.51955 -25.307344,49.25319 -48.688494,37.86649 -75.6645,-36.84884 -154.82438,3.95792 -183.70386,48.91322 -24.6677,38.39902 -103.40848,38.86978 -109.13362,1.70236 0,0 15.87125,-42.97327 15.04579,-46.55297 -19.35753,-83.94612 28.04205,-143.624812 0.31533,-188.1325516 z",
          "id": "path187",
          "label": "#Społeczeństwo",
          "style": {
            "clipRule": "evenodd",
            "display": "inline",
            "fill": "#a3d293",
            "fillOpacity": "1",
            "fillRule": "evenodd",
            "imageRendering": "optimizeQuality",
            "paintOrder": "stroke fill markers",
            "shapeRendering": "geometricPrecision",
            "stroke": "none",
            "strokeDasharray": "none",
            "strokeLinecap": "round",
            "strokeLinejoin": "round",
            "strokeOpacity": "1",
            "strokeWidth": "1",
            "textRendering": "geometricPrecision",
          },
        },
      },
    ],
  },
  "layer2": {
    "attributes": {
      "id": "layer2",
      "label": "L2",
      "style": {
        "display": "inline",
      },
    },
    "children": [
      {
        "path": {
          "boundingBox": {
            "center": {
              "x": -390.95909444203926,
              "y": 132.81968177008088,
            },
            "max": {
              "x": -318.7710632026188,
              "y": 247.13325685152185,
            },
            "min": {
              "x": -463.14712568145967,
              "y": 18.506106688639907,
            },
          },
          "d": "m -363.4396,241.68489 c 3.96619,6.14182 16.48527,7.44305 21.04765,2.08846 2.4872,-2.91897 4.71538,-4.54032 5.94819,-10.52523 3.03585,-14.7382 1.34456,-11.17449 1.82763,-25.49292 0.76914,-22.80346 20.75208,-29.3865 14.70798,-56.26046 -2.76562,-12.29581 -21.23925,-26.57881 -25.61829,-37.5256 -8.34906,-20.871457 -9.56328,-37.659787 -21.78509,-56.279572 -4.08805,-6.227937 -7.36626,-13.710207 -12.69914,-18.040411 -6.4217,-5.21482 -14.88743,-12.647945 -21.7881,-14.977616 -2.41944,-0.816986 -7.06024,-4.418125 -10.48298,-5.395171 -6.8687,-1.960855 -27.78301,-0.320121 -33.75197,6.566544 -7.15806,8.258677 -12.73002,18.888957 -15.49318,32.684379 -4.77058,23.817021 1.32816,32.000403 13.90025,45.992407 16.43391,18.28929 30.3,30.36147 42.98876,57.21694 3.23503,6.84699 11.60076,28.70405 14.07047,37.68774 1.29374,4.70623 5.98891,16.04619 8.35371,19.39709 3.03448,4.3009 8.53048,13.45997 12.0759,16.23376 0.49509,0.3873 6.59438,6.4687 6.69821,6.62966 z",
          "id": "path1",
          "label": "#Komputery i komunikacja",
          "style": {
            "clipRule": "evenodd",
            "display": "inline",
            "fill": "#bbe9ab",
            "fillOpacity": "1",
            "fillRule": "evenodd",
            "imageRendering": "optimizeQuality",
            "shapeRendering": "geometricPrecision",
            "stroke": "none",
            "strokeDasharray": "none",
            "strokeOpacity": "0.609571",
            "strokeWidth": "0.2",
            "textRendering": "geometricPrecision",
          },
        },
      },
      {
        "path": {
          "boundingBox": {
            "center": {
              "x": 220.04397547889005,
              "y": 84.34425781473227,
            },
            "max": {
              "x": 288.4188627868724,
              "y": 220.4664682145745,
            },
            "min": {
              "x": 151.66908817090766,
              "y": -51.77795258510995,
            },
          },
          "d": "m 212.20335,-34.669026 c -8.94654,-4.860515 -8.34878,-13.962352 -20.30352,-17.040296 -6.83225,-1.588798 -7.73113,24.862583 -9.44043,32.185241 -0.91912,3.589416 -12.60608,53.335355 -14.3674,63.103834 1.38994,17.368799 45.26621,29.789013 37.65977,56.15418 -8.1943,28.134337 -49.42526,75.315347 -51.96557,85.204967 -7.71153,30.02157 7.2021,30.62542 19.68907,34.07569 23.63115,9.96491 24.34655,-34.01855 31.88103,-51.09838 4.65662,-10.49641 23.04265,-48.55132 26.07288,-52.25861 12.19877,-12.18611 17.02906,-33.110125 34.2917,-42.815738 9.93192,-6.414989 28.04278,-37.514535 21.18502,-48.544315 -3.01256,-9.444589 -27.73993,-28.7133653 -36.27902,-34.843559 -11.07141,-4.713332 -28.18296,-16.820852 -38.42353,-24.123014 z",
          "id": "path2",
          "label": "#Bakterie i choroby zakaźne",
          "style": {
            "clipRule": "evenodd",
            "display": "inline",
            "fill": "#bbe9ab",
            "fillOpacity": "1",
            "fillRule": "evenodd",
            "imageRendering": "optimizeQuality",
            "shapeRendering": "geometricPrecision",
            "stroke": "none",
            "strokeDasharray": "none",
            "strokeOpacity": "0.609571",
            "strokeWidth": "0.160109",
            "textRendering": "geometricPrecision",
          },
        },
      },
      {
        "path": {
          "boundingBox": {
            "center": {
              "x": -288.67674837482605,
              "y": 224.74662880840003,
            },
            "max": {
              "x": -244.87114689618403,
              "y": 257.35257682356297,
            },
            "min": {
              "x": -332.48234985346807,
              "y": 192.1406807932371,
            },
          },
          "d": "m -246.11542,228.15383 c -1.79211,-5.90935 -10.29418,-17.6088 -15.19891,-21.42488 -14.40245,-11.2056 -34.01127,-16.91207 -50.82579,-13.69917 -7.23073,1.38147 -16.20355,8.50804 -18.90825,17.28622 -2.00358,6.50259 -1.66977,14.36515 -0.44871,21.52437 3.1047,18.20272 22.40948,23.61241 34.66899,24.66315 9.22476,0.79064 17.99186,1.09154 27.24749,0.63402 3.02328,-0.14942 14.55217,-7.04433 16.70963,-8.50253 5.9681,-4.03374 10.14222,-13.16397 6.82932,-19.71591 -0.076,-0.1501 -0.0355,-0.63915 -0.0738,-0.76527 z",
          "id": "path4",
          "label": "#Tekst i tłumaczenie",
          "style": {
            "clipRule": "evenodd",
            "display": "inline",
            "fill": "#bbe9ab",
            "fillOpacity": "1",
            "fillRule": "evenodd",
            "imageRendering": "optimizeQuality",
            "shapeRendering": "geometricPrecision",
            "stroke": "none",
            "strokeDasharray": "none",
            "strokeOpacity": "0.609571",
            "strokeWidth": "0.174368",
            "textRendering": "geometricPrecision",
          },
        },
      },
    ],
  },
  "layer3": {
    "attributes": {
      "id": "layer3",
      "label": "L3",
      "style": {
        "display": "inline",
      },
    },
    "groups": [
      {
        "attributes": {
          "id": "layer4",
          "label": "L3-maszyny",
        },
        "children": [
          {
            "rect": {
              "boundingBox": {
                "center": {
                  "x": -319.7775159,
                  "y": -214.0745055,
                },
                "max": {
                  "x": -316.76377180000003,
                  "y": -210.98814099999998,
                },
                "min": {
                  "x": -322.79126,
                  "y": -217.16087,
                },
              },
              "height": 6.172729,
              "id": "rect31",
              "label": "#Urządzenia mechaniczne",
              "style": {
                "display": "inline",
                "fill": "#000000",
              },
              "width": 6.0274882,
              "x": -322.79126,
              "y": -217.16087,
            },
          },
          {
            "rect": {
              "boundingBox": {
                "center": {
                  "x": -297.8461759,
                  "y": -230.9224255,
                },
                "max": {
                  "x": -294.8324318,
                  "y": -227.836061,
                },
                "min": {
                  "x": -300.85992,
                  "y": -234.00879,
                },
              },
              "height": 6.172729,
              "id": "rect27",
              "label": "#Robotyka",
              "style": {
                "display": "inline",
                "fill": "#000000",
              },
              "width": 6.0274882,
              "x": -300.85992,
              "y": -234.00879,
            },
          },
        ],
      },
      {
        "attributes": {
          "id": "layer6",
          "label": "L3-Ziemia",
        },
        "children": [
          {
            "rect": {
              "boundingBox": {
                "center": {
                  "x": -141.0597459,
                  "y": -293.0270355,
                },
                "max": {
                  "x": -138.0460018,
                  "y": -289.940671,
                },
                "min": {
                  "x": -144.07349,
                  "y": -296.1134,
                },
              },
              "height": 6.172729,
              "id": "rect32",
              "label": "#Asfalt",
              "style": {
                "display": "inline",
                "fill": "#000000",
              },
              "width": 6.0274882,
              "x": -144.07349,
              "y": -296.1134,
            },
          },
          {
            "rect": {
              "boundingBox": {
                "center": {
                  "x": 6.667645,
                  "y": -297.6626255,
                },
                "max": {
                  "x": 9.6813891,
                  "y": -294.576261,
                },
                "min": {
                  "x": 3.6539009,
                  "y": -300.74899,
                },
              },
              "height": 6.172729,
              "id": "rect33",
              "label": "#Wydobywanie węglowodorów",
              "style": {
                "display": "inline",
                "fill": "#000000",
              },
              "width": 6.0274882,
              "x": 3.6539009,
              "y": -300.74899,
            },
          },
        ],
      },
    ],
  },
}
`;
